rootston/output.c

```diff
@@ -68,22 +81,41 @@ static void render_surface(struct wlr_surface *surface,
 		int ow, oh;
 		wlr_output_effective_resolution(wlr_output, &ow, &oh);
 
-		double c = cos(rotation);
-		double s = sin(rotation);
-
-		// Odd transforms are rotations
-		double r = wlr_output->transform % 2 == 0 ? 1 : -1;
-
 		float mat[9] = {
-			+r * (sw * c)/ow, +r * (sh * -s)/ow, r * ((+2.0f * ox + sw) / ow - 1.0f),
-			-r * (sw * s)/oh, -r * (sh * +c)/oh, r * ((-2.0f * oy - sh) / oh + 1.0f),
+			cos(rotation), -sin(rotation), 0.0f,
+			sin(rotation), +cos(rotation), 0.0f,
+			0.0f, 0.0f, 1.0f,
+		};
+
+		// Scale matrix to window size
+		float scale[9] = {
+			sw,   0.0f, 0.0f,
+			0.0f, sh,   0.0f,
+			0.0f, 0.0f, 1.0f,
+		};
```

Comment:
I liked the conveinence functions for creating rotation/scale/translation matricies, too. Was there a particular reason you left them out?

Reply:
Basically, they were unneeded. I'm deliberately trying to keep the API small.
It was also kind of an inefficient API.

Reply:
I think they were pretty small to begin with, and pretty convenient. It's a lot better design-wise to be able to explicitly say `get_rotation_matrix(float radians)` or something to that effect, as opposed to having to remember how to construct a rotation matrix (I'd have to look it up).

So far as effeciency is concerned, wouldn't the approach of only recalculating the matricies when views are invalidated address that issue?

Reply:
It's really something that I don't think adds a lot of value. It's something which isn't core to wlroots AT ALL (as in, it makes no use of any library internals), and you can find dozens of more efficient single header C "libraries" which can do matrix manipulation extremely easily.

Maybe the functions could live in rootston, but I really don't think we should be bloating the wlroots API with that kind of stuff.

Reply:
I am very opposed to the idea of "single header C 'libraries'". I think they're an awful idea and they subvert a lot of good C programming principles. They might be "more effecient", but our functions are *well* within our performance budget. Even on the highest end of displays our bottleneck is the display's framerate. I'm sure we could compute our matricies tens of thousands of times between frames.

When it comes to API bloat, I don't think matricies really count. The functions are small in number and size. Matrix math is also not a controversial API - it's simple mathematics. We aren't making any design decisions on the user's behalf by providing such an API. We already have precedent for this sort of thing, too - the convenience functions we offer for wlr_box are something we could easily avoid exporting, but they're simple to implement and something many compositors will need.

Reply:
Having some util code that we use internally (like wlr_box) which we export is different. It's going to be there anyway, so you may as well make it part of the API.

However, matricies are not used anywhere internally at all, and I don't think it will ever be needed (obviously excluding the output rotations internal to just the renderer).
I consider this kind of code to be completely superfluous. It basically ties into the same reason I hate wlr_list so much. Sure, lists are useful and all, but I don't think it's our job to be providing them over our API.
If we start providing excess util code, where does it end? Should we provide a hashmap too? Those are pretty useful.

I don't even consider matricies to be common use case. They only there because having support for window rotation was insisted, but it's really just for gimmicky demo features.  I think most realistic uses of wlroots would't even need this.

Reply:
Many of the things we provide are for convenience. If we didn't do that, we'd just be left with unopinionated state machines that implement Wayland interfaces. Tools like wlr_output_layout, wlr_cursor, and wlr_xcursor would not exist. You could even make an argument that our backends wouldn't exist, too. We provide tools that we think compositors will need, but don't couple them so tightly that they can't go without. Matricies are one such tool.

Rotation might seem like a gimmick to you, but only on compositors like Sway. Remember that not everyone prefers a minimalistic experience. Many other kinds of compositors exist, and it's our job to design APIs that suit the broadest audience of compositors. Imagine all of the more novel matrix transformations that would be applied on a less gimmicky compositor that uses animations? Do you use android? Open the app switcher and pull it down beyond the maximum extent. Note the shearing and use of perspective here - is this a gimmick? Mobile compositors are something wlroots is intended to support in addition to desktop compositors. Check out some of GNOME or KDE's animations, too - not all of them can be accomplished by simply mapping the surface to a rectangle on-screen.

I don't like wlr_list either, but it's there because it's the most suited data structure to some of our API and we need to expose its functions so those parts of our API can be interacted with effectively. We could probably make a separate issue about making most of its functions internal, though.

Reply:
To go back to what I said earlier
>This was only designed for pretty basic 2D compositing, about to the requirements of what Sway or something like Openbox would need

I'm honestly not interested in supporting anything beyond that with wlr_render, for the reasons I described in one of my previous posts. "render_texture_with_matrix" just exists to subvert the interface.

>Matrix math is also not a controversial API - it's simple mathematics

That's part of the reason I'm against it. Since it's so simple, why even bother having it?
The user can just throw together 30 lines of code with their own matrix stuff.
There aren't 500 lines of code which interacts with other systems and would be considered essential for traditional window managers.

Reply:
I'd add neither, the matrix interface to rendering nor the helper functions, because I do not see a practical use case. But if you're adding a matrix interface, helper functions to build matrices in the right format should be there as well.

If you were to use them for animations, wouldn't a simple matrix-based transformation be too simple to get anything worthwhile anyway?

Reply:
>If you were to use them for animations, wouldn't a simple matrix-based transformation be too simple to get anything worthwhile anyway?

Yes, that's basically my original point from a couple of weeks ago. This renderer is (intentionally) far too limited to do any of the special use-cases being discussed.

Reply:
Alright, whatever. You can revert the matrix addition and remove the rotation code from rootston.

Reply:
It's already there. It shares like 90% of the code with the "normal" interface, so it's not adding a huge amount of code.
I don't really think it needs to be there, but it's not something that I feel **that** strongly about.

Reply:
Either you want matricies or you don't commit to one and fucking do it right.